File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/infrastructure/PaymentAccountAdapterImpl.java
package com.example.accountservice.payment.infrastructure;

import com.example.accountservice.account.application.query.AccountQuery;
import com.example.accountservice.account.domain.Account;
import com.example.accountservice.account.domain.AccountFindQuery;
import com.example.accountservice.account.domain.AccountStatus;
import com.example.accountservice.payment.application.adapter.AccountAdapter;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

/**
 * The implementation (ACL) of {@link AccountAdapter}. Injects and calls the read interface ({@link
 * AccountQuery}) exposed by the Account BC, and translates the Account BC's model into the minimal
 * shape the Payment BC uses ({@link AccountAdapter.AccountView}). Unlike Card BC's
 * AccountAdapterImpl, this also needs to translate the balance (balanceAmount), since the balance
 * is needed to judge whether a payment can proceed.
 */
// Why the class is named PaymentAccountAdapterImpl: Spring's default bean name is derived from
// the simple class name, so if it shared a name with card/infrastructure/AccountAdapterImpl
// (owned by Card BC), component scanning would throw ConflictingBeanDefinitionException — bean
// names must be globally unique even across different packages.
@Component
@RequiredArgsConstructor
public class PaymentAccountAdapterImpl implements AccountAdapter {

    private final AccountQuery accountQuery;

    @Override
    public Optional<AccountView> findAccount(String accountId, String ownerId) {
        return accountQuery
                .findAccounts(new AccountFindQuery(0, 1, accountId, ownerId, null))
                .accounts()
                .stream()
                .findFirst()
                .map(this::toAccountView);
    }

    private AccountView toAccountView(Account account) {
        return new AccountView(
                account.getAccountId(),
                account.getStatus() == AccountStatus.ACTIVE,
                account.getBalance().amount(),
                account.getBalance().currency());
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/infrastructure/PaymentAccountAdapterImpl.kt
package com.example.accountservice.payment.infrastructure

import com.example.accountservice.account.application.query.AccountQuery
import com.example.accountservice.account.domain.AccountFindQuery
import com.example.accountservice.account.domain.AccountStatus
import com.example.accountservice.payment.application.adapter.AccountAdapter
import com.example.accountservice.payment.application.adapter.AccountView
import org.springframework.stereotype.Component

/**
 * The implementation of [AccountAdapter] (ACL). Injects and calls the read port ([AccountQuery])
 * exposed by Account BC, and translates Account BC's model into the minimal shape Payment BC uses
 * ([AccountView]) — including the balance (balanceAmount), moving it into the shape needed to judge
 * "whether payment is possible."
 *
 * Why the class name carries a `Payment` prefix: Card BC already has
 * [com.example.accountservice.card.infrastructure.AccountAdapterImpl] playing the same role, and
 * Spring's default bean-name generation looks only at the simple class name (even across different
 * packages), so two `AccountAdapterImpl` classes trigger a `ConflictingBeanDefinitionException` — this
 * naming convention exists specifically to avoid that collision.
 */
@Component
class PaymentAccountAdapterImpl(
    private val accountQuery: AccountQuery,
) : AccountAdapter {
    override fun findAccount(
        accountId: String,
        ownerId: String,
    ): AccountView? {
        val (accounts, _) =
            accountQuery.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = accountId, ownerId = ownerId))
        return accounts.firstOrNull()?.let { account ->
            AccountView(
                accountId = account.accountId,
                active = account.status == AccountStatus.ACTIVE,
                balanceAmount = account.balance.amount,
                currency = account.balance.currency,
            )
        }
    }
}
